File: legacy/src/game/pinball-knight/render/ingest/slice.ts
/**
 * SLICING — a matted sheet into rows of cells.
 *
 * Extracted verbatim from `render/sprite-ingest.test.ts`, where it lived inside
 * the vitest file and therefore could only ever run in node. The browser refiner
 * (`/sprites`) needs the same code, and a re-implementation would drift the first
 * time a threshold moved — the same reasoning that killed the Python importer's
 * private copy of the palette.
 *
 * Everything here is pure: `Uint8ClampedArray` in, plain coordinates out. No
 * canvas, no filesystem, no DOM. That is what lets it run in vitest AND in the
 * browser without a shim.
 *
 * ⚠️ THIS FINDS CELLS BY ALPHA. It assumes matting already happened — see
 * `matte.ts`. Handed a sheet on an opaque background it returns ONE cell, which
 * is what the caller's guard is for.
 */

/** Transparent columns/rows narrower than this do not separate two cells. */
export const MIN_GAP = 6;
/** A run of opaque pixels smaller than this is a smudge, not a pose. */
export const MIN_CELL = 12;

/**
 * A row or column this full of opaque pixels is a RULED GRID LINE, not art.
 *
 * Real sheets are drawn with cell borders. They are opaque, so a naive
 * alpha-slice sees the whole sheet as ONE connected region and returns a single
 * cell — which is why this exists rather than being defensive programming.
 * Art never spans a whole sheet dimension; a ruled line always does.
 *
 * ⚠️ THE SECOND SENTENCE IS FALSE AS APPLIED, and it is the bug. Measured on
 * `fixtures.ts` (see slice.test.ts):
 *
 *   · FILL IS NOT ENOUGH — a rule is also THIN. A figure's own solid core is a
 *     column of opaque pixels spanning the whole band, so this test erases it.
 *     On an unruled sheet, 176 of 176 opaque columns were stripped: the figures
 *     vanished and the sheet sliced to NOTHING. The shipped comments blame the
 *     art for "splitting inside a figure wherever a pose leaves a transparent
 *     column"; the slicer manufactures that column itself.
 *
 *   · THE DENOMINATOR IS WRONG — fill is measured against the SHEET width, but
 *     a ragged sheet's short rows are nowhere near it. A 4-cell row inside a
 *     6-cell sheet fills 56%, so its borders survive and weld the row into one
 *     cell. That is the documented "1/6/1/1/1", and it reproduces exactly.
 *
 * Requiring a rule to be tall AND ≤3px wide, and measuring fill against the
 * row's own extent, slices the unruled fixture to exactly 4/6/4/2/3. `grid.ts`
 * carries that fix; this constant is left as-is so the characterisation tests
 * keep describing what shipped.
 */
export const RULE_FILL = 0.7;

/**
 * Bands shorter than this fraction of the median band are CAPTIONS.
 *
 * Sheets label their rows ("IDLE", "SPRING ATTACK", "DEATH"). The lettering
 * sits on the background between rows, so it slices as its own short band and
 * would otherwise be imported as a pose. A caption is an order of magnitude
 * shorter than a figure; 0.25 separates them with room to spare.
 *
 * ⚠️ THIS ONLY CATCHES CAPTIONS THAT SIT IN THEIR OWN BAND — i.e. above or below
 * a row. A caption in a LEFT GUTTER shares the row's band, so its height is the
 * row's height and this test never sees it; it then survives the width filter in
 * `sliceSheet` (a 6-letter word is well over a quarter of a cell) and imports as
 * a frame. Gutter labels have to be stripped before ingest, or declared as a
 * left margin in the recipe.
 */
export const CAPTION_RATIO = 0.25;

/** One cell, as `[x0, y0, x1, y1]` inclusive. */
export type Cell = [number, number, number, number];

export interface SheetRow {
  cells: Cell[];
}

/** Contiguous true runs, merging gaps under MIN_GAP and dropping tiny runs. */
export function bands(profile: boolean[]): [number, number][] {
  const out: [number, number][] = [];
  let start = -1;
  let gap = 0;
  for (let i = 0; i < profile.length; i++) {
    if (profile[i]) {
      if (start < 0) start = i;
      gap = 0;
    } else if (start >= 0 && ++gap > MIN_GAP) {
      if (i - gap - start >= MIN_CELL) out.push([start, i - gap]);
      start = -1;
    }
  }
  if (start >= 0 && profile.length - start >= MIN_CELL) out.push([start, profile.length - 1]);
  return out;
}

/**
 * Slice a sheet into ROWS of cells, discarding ruled grid lines and captions.
 *
 * Returns rows rather than a flat list because real sheets are ragged — one
 * observed sheet runs 4 / 6 / 4 / 2 / 3 across its five clips, and a row that
 * does not start at column 1. Flattening first would make "which frame belongs
 * to which clip" unrecoverable.
 */
export function sliceSheet(data: Uint8ClampedArray, w: number, h: number): SheetRow[] {
  const solid = new Uint8Array(w * h);
  for (let i = 0; i < w * h; i++) solid[i] = data[i * 4 + 3] > 8 ? 1 : 0;

  // ── Strip ruled lines. A full-width row or full-height column of opaque
  // pixels is a border; art never spans the sheet.
  for (let y = 0; y < h; y++) {
    let n = 0;
    for (let x = 0; x < w; x++) n += solid[y * w + x];
    if (n >= w * RULE_FILL) for (let x = 0; x < w; x++) solid[y * w + x] = 0;
  }
  // ⚠️ VERTICAL RULES ARE STRIPPED PER-BAND, NOT SHEET-WIDE, and that is the
  // whole difference between this working and not. A cell border is only as
  // tall as ITS ROW, so against the full sheet height a 5-row sheet's border
  // fills about 18% — nowhere near any sensible threshold — and survives. It
  // then bridges the gap between neighbouring cells and the entire row slices as
  // ONE frame. Measured on a fixture with ruled cells: sheet-wide stripping gave
  // rows of 1/6/1/1/1 where the truth was 4/6/4/2/3.

  const at = (x: number, y: number): boolean => solid[y * w + x] === 1;
  const rowProfile: boolean[] = [];
  for (let y = 0; y < h; y++) {
    let any = false;
    for (let x = 0; x < w && !any; x++) any = at(x, y);
    rowProfile.push(any);
  }

  // ── Drop caption bands, by height against the median band.
  const raw = bands(rowProfile);
  const heights = raw.map(([a, b]) => b - a + 1).sort((a, b) => a - b);
  const median = heights[Math.floor(heights.length / 2)] ?? 0;
  const keep = raw.filter(([a, b]) => b - a + 1 >= median * CAPTION_RATIO);

  const out: SheetRow[] = [];
  for (const [y0, y1] of keep) {
    const bandH = y1 - y0 + 1;
    // Band-local copy with rules stripped on BOTH axes. A cell's top and bottom
    // edges span its full width, so leaving them in makes every column look like
    // content and welds the row back together.
    const band = new Uint8Array(w * bandH);
    for (let y = 0; y < bandH; y++) for (let x = 0; x < w; x++) band[y * w + x] = at(x, y0 + y) ? 1 : 0;
    for (let y = 0; y < bandH; y++) {
      let n = 0;
      for (let x = 0; x < w; x++) n += band[y * w + x];
      if (n >= w * RULE_FILL) for (let x = 0; x < w; x++) band[y * w + x] = 0;
    }
    for (let x = 0; x < w; x++) {
      let n = 0;
      for (let y = 0; y < bandH; y++) n += band[y * w + x];
      if (n >= bandH * RULE_FILL) for (let y = 0; y < bandH; y++) band[y * w + x] = 0;
    }
    const inBand = (x: number, y: number): boolean => band[(y - y0) * w + x] === 1;
    const colProfile: boolean[] = [];
    for (let x = 0; x < w; x++) {
      let any = false;
      for (let y = y0; y <= y1 && !any; y++) any = inBand(x, y);
      colProfile.push(any);
    }
    const cells: [number, number, number, number, number][] = [];
    for (const [x0, x1] of bands(colProfile)) {
      // Tighten vertically to this cell's own ink — the band is the union
      // across the row, and a crouched pose is shorter than its neighbours.
      let ty = y1, by = y0, mass = 0;
      for (let y = y0; y <= y1; y++) {
        for (let x = x0; x <= x1; x++) {
          if (!inBand(x, y)) continue;
          mass++;
          ty = Math.min(ty, y);
          by = Math.max(by, y);
        }
      }
      if (by >= ty) cells.push([x0, ty, x1, by, mass]);
    }
    // Reject FRAGMENTS by WIDTH against the row's median cell.
    //
    // Mass is the wrong test and was tried first: a leftover ruled border is
    // long, so it carries real mass (a 2x260 edge is 520 px) and survived a 2%
    // threshold, then crushed to 8 texels and tripped the empty-cell guard.
    // Width is the discriminator — a border remnant is a couple of pixels wide
    // where a pose is a couple of hundred — and it stays correct for a small
    // pose, which a mass test does not: a death sprawl is legitimately light.
    const widths = cells.map((c) => c[2] - c[0] + 1).sort((a, b) => a - b);
    const medianW = widths[Math.floor(widths.length / 2)];
    const real = cells
      .filter((c) => c[2] - c[0] + 1 >= medianW * 0.25 && c[4] > 0)
      .map((c) => [c[0], c[1], c[2], c[3]] as Cell);
    if (real.length) out.push({ cells: real });
  }
  return out;
}

/**
 * Re-cut a band into exactly `n` equal columns across its own opaque extent.
 *
 * ⚠️ THIS EXISTS BECAUSE AUTO-SLICING A RULED SHEET DOES NOT WORK, and pretending
 * otherwise would be worse than asking for one number. Measured on a fixture
 * built to match a real supplied sheet (ruled cells, ragged rows 4/6/4/2/3, one
 * row indented), alpha-slicing returned 5/12/5/2/1: it splits on the border
 * remnants, splits AGAIN inside a figure wherever a pose leaves a transparent
 * column — between the legs, either side of a spring — and merges neighbours
 * whose art touches. Those three failure modes pull in opposite directions, so
 * no gap threshold fixes all of them.
 *
 * Cells on a real sheet are laid out on a regular pitch, so given the count, an
 * equal division across the band's extent is exact. An indented row divides
 * correctly too, because the extent is the row's own bounding box, not the
 * sheet's.
 *
 * ⚠️ IT DIVIDES THE INK EXTENT, NOT THE CELL EXTENT. A row whose first pose is
 * narrow and inset, or whose last pose stops short of its cell, has an ink
 * extent smaller than its true grid extent — so the pitch is slightly wrong and
 * the error ACCUMULATES across the row. `grid.ts` exists to remove this: the
 * ruled borders are the true pitch, and the ink is not.
 */
export function equalCells(band: SheetRow, n: number): Cell[] {
  const x0 = Math.min(...band.cells.map((c) => c[0]));
  const x1 = Math.max(...band.cells.map((c) => c[2]));
  const y0 = Math.min(...band.cells.map((c) => c[1]));
  const y1 = Math.max(...band.cells.map((c) => c[3]));
  const pitch = (x1 - x0 + 1) / n;
  return Array.from({ length: n }, (_, i) => [
    Math.round(x0 + i * pitch), y0, Math.round(x0 + (i + 1) * pitch) - 1, y1,
  ] as Cell);
}

File: ThreeJS/src/game/pinball-knight/tools/sprite-forge/python/spriteforge/slice.py
"""
SLICING — a matted sheet into ragged rows of cells.

Returns ROWS rather than a flat list because real sheets are ragged: one observed
sheet runs 4 / 6 / 4 / 2 / 3 across its five clips, with a row that does not start
at column 1. Flattening first makes "which frame belongs to which clip"
unrecoverable.

⚠️ THIS FINDS CELLS BY ALPHA. It assumes matting already happened. Handed a sheet
on an opaque background it returns ONE cell, which is what the caller's guard is
for.

── WHAT A RULED LINE IS, AND THE TWO WAYS THAT WAS GOT WRONG ───────────────

Real sheets are drawn with cell borders. They are opaque, so a naive alpha-slice
sees the whole sheet as ONE region and returns a single cell.

FILL ALONE IS NOT THE TEST, and believing it was cost this module its reputation
twice:

  · A RULE IS ALSO THIN. A figure's own solid core is a column of opaque pixels
    spanning the whole band, so a height-only test erases it. On an unruled sheet
    176 of 176 opaque columns were stripped and the sheet sliced to NOTHING.

  · A RULE IS ALSO CONTIGUOUS. Total ink counts a row of separate figures the
    same as a line drawn through them, so a sheet of BROAD creatures busts the
    threshold on the strength of the creatures alone. Measured on a frog sheet —
    five wide frogs per row — the widest scanlines reach 73% total ink and were
    erased as borders, so the idle cells came back 57px tall against a ~150px
    frog and every frame shipped as a headless dome. Longest CONTIGUOUS run
    separates them by construction:

        a ruled border   ~100% contiguous
        the frog row       15% contiguous   (73% total ink)
"""

from __future__ import annotations

from dataclasses import dataclass

import numpy as np

#: Transparent columns/rows narrower than this do not separate two cells.
MIN_GAP = 6
#: A run of opaque pixels smaller than this is a smudge, not a pose.
MIN_CELL = 12
#: A row or column this CONTIGUOUSLY full of opaque pixels is a ruled grid line.
RULE_FILL = 0.7
#: A vertical rule may be at most this share of the band's HEIGHT in width.
#:
#: Scale-free on purpose. An absolute "3px" is a guess about the source
#: resolution and breaks the first time a sheet arrives upscaled 4x with 8px
#: borders. Against the band, a ruled border is 1-3px of a ~120px cell (1-2.5%)
#: and a figure's core is 44px (37%) — a 15x margin either side.
RULE_MAX_W = 0.1
#: ...but never below this, so a short band cannot make the cap sub-pixel.
RULE_MIN_W = 3
#: Bands shorter than this share of the median band are CAPTIONS.
#:
#: Sheets label their rows ("IDLE", "SPRING ATTACK"). The lettering sits between
#: rows, slices as its own short band, and would import as a pose. A caption is an
#: order of magnitude shorter than a figure.
#:
#: ⚠️ This only catches captions in their OWN band. One in a LEFT GUTTER shares
#: the row's band, so its height is the row's height and this never sees it.
CAPTION_RATIO = 0.25

#: x0, y0, x1, y1 — inclusive.
Cell = tuple[int, int, int, int]


@dataclass
class SheetRow:
    cells: list[Cell]


@dataclass
class SliceDetail:
    rows: list[SheetRow]
    #: `1` where the sheet has art, at sheet coordinates.
    #:
    #: The slicer's OWN view: alpha keyed, sheet rules stripped, outer frame
    #: stripped, per-band cell borders stripped. The gutter check must measure
    #: THIS and not raw alpha, or a ruled sheet fails on its own borders — a 2px
    #: rule inside a 15px probe reads 13%.
    mask: np.ndarray


def bands(profile: np.ndarray) -> list[tuple[int, int]]:
    """Contiguous true runs, merging gaps under MIN_GAP and dropping tiny runs."""
    out: list[tuple[int, int]] = []
    start = -1
    gap = 0
    for i, v in enumerate(profile):
        if v:
            if start < 0:
                start = i
            gap = 0
        elif start >= 0:
            gap += 1
            if gap > MIN_GAP:
                if i - gap - start >= MIN_CELL:
                    out.append((start, i - gap))
                start = -1
    if start >= 0 and len(profile) - start >= MIN_CELL:
        out.append((start, len(profile) - 1))
    return out


def _longest_run(row: np.ndarray) -> int:
    """Longest contiguous run of 1s. A rule is a LINE, not scattered ink."""
    if not row.any():
        return 0
    # Run-length via the positions where the value changes.
    idx = np.flatnonzero(np.diff(np.concatenate(([0], row, [0]))))
    return int((idx[1::2] - idx[0::2]).max())


def slice_sheet_detail(alpha: np.ndarray) -> SliceDetail:
    """Slice a matted sheet. `alpha` is the HxW alpha channel."""
    h, w = alpha.shape
    solid = (alpha > 8).astype(np.uint8)

    # ── Sheet-wide horizontal rules, measured as ONE CONTIGUOUS RUN.
    for y in range(h):
        if _longest_run(solid[y]) >= w * RULE_FILL:
            solid[y] = 0

    # ── The OUTER frame's sides, and only those.
    #
    # A frame around the whole sheet puts ink in every row, so the row profile
    # never breaks and the entire sheet reads as ONE band. Safe here and
    # catastrophic per-band because of the threshold: an outer frame spans ~100%
    # of the sheet height where a CELL border in a 5-row sheet spans about 18%.
    frame_cap = max(RULE_MIN_W, h * 0.01)
    full_h = solid.sum(axis=0) >= h * 0.98
    x = 0
    while x < w:
        if not full_h[x]:
            x += 1
            continue
        end = x
        while end + 1 < w and full_h[end + 1]:
            end += 1
        if end - x + 1 <= frame_cap:
            solid[:, x : end + 1] = 0
        x = end + 1

    # ⚠️ CELL BORDERS ARE STRIPPED PER-BAND, NOT SHEET-WIDE. A cell border is only
    # as tall as ITS ROW, so against the full sheet height a 5-row sheet's border
    # fills about 18% — nowhere near any sensible threshold — and survives, then
    # bridges neighbouring cells so the row slices as ONE frame.
    row_profile = solid.any(axis=1)
    raw = bands(row_profile)
    if not raw:
        return SliceDetail(rows=[], mask=np.zeros((h, w), np.uint8))
    heights = sorted(b - a + 1 for a, b in raw)
    median = heights[len(heights) // 2]
    keep = [(a, b) for a, b in raw if b - a + 1 >= median * CAPTION_RATIO]

    mask = np.zeros((h, w), np.uint8)
    out: list[SheetRow] = []
    for y0, y1 in keep:
        band = solid[y0 : y1 + 1].copy()
        band_h = band.shape[0]

        # Fill is measured against THIS ROW's extent, not the sheet's. A ragged
        # sheet's short rows never reach 70% of the sheet width, so their borders
        # used to survive and weld the row into one cell.
        cols = np.flatnonzero(band.any(axis=0))
        if cols.size == 0:
            continue
        row_w = int(cols[-1] - cols[0] + 1)

        ruled = False
        for y in range(band_h):
            if _longest_run(band[y]) < row_w * RULE_FILL:
                continue
            band[y] = 0
            ruled = True

        # ── A CELL BORDER IS A RECTANGLE: no vertical rules without horizontal
        # ones. On an UNRULED sheet any column the vertical test flags is art by
        # construction — it flagged all of it, 176 of 176, and sliced to nothing.
        if ruled:
            # A vertical rule is SPANNING *AND* NARROW, and opaque at BOTH band
            # edges: a border runs the full height of its cell where art is inset.
            cap_w = max(RULE_MIN_W, band_h * RULE_MAX_W)
            tall = (band.sum(axis=0) >= band_h * RULE_FILL) & (band[0] == 1) & (band[-1] == 1)
            x = 0
            while x < w:
                if not tall[x]:
                    x += 1
                    continue
                end = x
                while end + 1 < w and tall[end + 1]:
                    end += 1
                # Cleared as RUNS, not columns: two neighbouring cells' borders
                # touch and read as one 2px rule, which is still a rule.
                if end - x + 1 <= cap_w:
                    band[:, x : end + 1] = 0
                x = end + 1

        mask[y0 : y1 + 1] |= band

        col_profile = band.any(axis=0)
        cells: list[tuple[int, int, int, int, int]] = []
        for cx0, cx1 in bands(col_profile):
            # Tighten vertically to this cell's own ink — the band is the union
            # across the row, and a crouched pose is shorter than its neighbours.
            sub = band[:, cx0 : cx1 + 1]
            ys = np.flatnonzero(sub.any(axis=1))
            if ys.size == 0:
                continue
            cells.append((cx0, y0 + int(ys[0]), cx1, y0 + int(ys[-1]), int(sub.sum())))

        if not cells:
            continue
        # Reject FRAGMENTS by WIDTH against the row's median cell.
        #
        # Mass is the wrong test and was tried first: a leftover ruled border is
        # long, so it carries real mass (a 2x260 edge is 520 px) and survived a 2%
        # threshold, then crushed to 8 texels and tripped the empty-cell guard.
        # Width is the discriminator, and it stays correct for a small pose, which
        # a mass test does not: a death sprawl is legitimately light.
        widths = sorted(c[2] - c[0] + 1 for c in cells)
        median_w = widths[len(widths) // 2]
        real = [(c[0], c[1], c[2], c[3]) for c in cells if c[2] - c[0] + 1 >= median_w * 0.25 and c[4] > 0]
        if real:
            out.append(SheetRow(cells=real))

    return SliceDetail(rows=out, mask=mask)


def slice_sheet(alpha: np.ndarray) -> list[SheetRow]:
    """`slice_sheet_detail`, rows only."""
    return slice_sheet_detail(alpha).rows


def equal_cells(row: SheetRow, n: int) -> list[Cell]:
    """Re-cut a row into exactly `n` equal columns across its own opaque extent.

    ⚠️ NOT THE NORMAL PATH, and it costs something. It divides the INK EXTENT,
    not the cell extent, so a row whose first pose is inset or whose last stops
    short drifts: measured on a frog sheet the true pitch was 266px and equal
    division gave 255, putting the fifth frame 47px off its figure. It also gives
    every cell the BAND's height rather than its own ink, which shrinks the figure.

    Kept for the sheet that still fools the alpha pass: two poses genuinely
    touching with no gap read as one cell, and no threshold recovers that from
    pixels alone.
    """
    x0 = min(c[0] for c in row.cells)
    x1 = max(c[2] for c in row.cells)
    y0 = min(c[1] for c in row.cells)
    y1 = max(c[3] for c in row.cells)
    step = (x1 - x0 + 1) / n
    return [(round(x0 + i * step), y0, round(x0 + (i + 1) * step) - 1, y1) for i in range(n)]
